### PR TITLE
feat(normalize): extensible provider registry

### LINKED ISSUE #5: feat: extensible provider registry (Gmail, Outlook, Yahoo rules)
## Summary

Replace hardcoded `is_gmail_domain()` with an extensible `ProviderRegistry`. Each provider defines: dot policy, case policy, subaddress separator, known aliases.

## Design

```rust
struct ProviderRule {
    domains: &[&str],           // ["gmail.com", "googlemail.com"]
    dot_policy: DotPolicy,      // Always strip
    case_policy: CasePolicy,    // All lowercase
    subaddress_sep: char,       // '+'
    is_freemail: bool,
}
```

Built-in providers: Gmail, Outlook/Hotmail, Yahoo, ProtonMail, iCloud, Yandex, Mail.ru.
User can extend: `Config::builder().add_provider(rule).build()`.

## Time estimate

4h

### PR DESCRIPTION
**Pull request by polaz**
## Summary

Replaces the hardcoded `is_gmail_domain()` and `is_freemail_domain()` lists with an extensible `ProviderRegistry` of `ProviderRule`s (Closes #5).

Each provider rule carries: domains, dot-stripping, case folding, subaddress separator, freemail flag. Applications can register their own.

## API

- **`ProviderRule::new(domains)`** + builder setters: `.strip_dots()`, `.lowercase_local()`, `.subaddress_separator(Option<char>)`, `.freemail()`.
- **`ProviderRegistry`** — `builtin()` / `empty()` / `add()` / `with()` / `lookup()`. Default = built-ins. User rules take precedence over built-ins (can redefine a built-in).
- **`Config::builder()`** — new `.provider_aware()`, `.add_provider(rule)`, `.providers(registry)`.

## Behavior

- **`provider_aware()` (opt-in):** when the domain matches a registered provider, that rule's dot/case/separator govern the address instead of the global policies. Non-matching domains keep the global policies. No surprise auto-normalization when off.
- **`is_freemail()`** now reads the registry (built-ins + custom), **always** — independent of `provider_aware`. Covers every domain the old hardcoded list recognized (verified by test).
- **`DotPolicy::GmailOnly`** consults only the **built-in** registry (Gmail/Googlemail) — custom providers added via `add_provider()` never affect this legacy mode; they apply only under `provider_aware()`. The built-in set lives in a process-wide `LazyLock`, so this hot path borrows it without a per-call allocation (matters for batch parsing).
- **Quoted local-parts are literal:** provider case-folding (and dots/subaddress) are skipped inside a quoted string; only a global lowercase policy still applies.

Built-in providers: Gmail/Googlemail (dot-stripping), Outlook/Hotmail/Live/MSN, Yahoo, ProtonMail, iCloud, Yandex, Mail.ru, plus AOL/Mail.com/Zoho/GMX/Web.de/Tutanota/Fastmail (freemail-only).

## Correctness fixes (caught in review)

- **GmailOnly leak:** custom `strip_dots` rules from `add_provider()` no longer leak into `DotPolicy::GmailOnly` when `provider_aware()` is off (regression test + fix).
- **Quoted-local case folding:** `"A.B"@gmail.com` under `provider_aware()` no longer folds to `a.b` (provider folding gated to unquoted locals; regression test + fix).
- IDN lookup path covered for Unicode NFC (decomposed vs composed) and confusable non-match.

## CI / tooling

- `structured-public-domains` is excluded from Dependabot — it is owned exclusively by the PSL data cascade (`psl-cascade.yml`), which cuts an off-tag data-only release. A Dependabot bump on main HEAD would route a PSL data update through the code path and race the cascade.

## Public API note

No removals — additive only. `DotPolicy::GmailOnly`, `dots_gmail_only()`, etc. all preserved (semver-friendly on the published 0.0.x crate).

## Testing

- `cargo nextest run --all-features` → 135 passed
- `cargo nextest run --no-default-features` → clean
- `cargo test --doc --all-features` → 6 passed
- `cargo clippy --all-features` and `--no-default-features` with `-D warnings` → clean
- `cargo fmt --all -- --check` → clean
- README updated with a Provider-Aware Normalization section.

Closes #5

## Summary by CodeRabbit

* **New Features**
  * Added optional provider-aware email normalization with a configurable provider registry (local-part case folding, dot stripping, and subaddress/tag separators).
  * Provider rules can be customized and override built-in behavior.
* **Documentation**
  * Updated README with a “Provider-Aware Normalization” section, including configuration examples and freemail/provider-specific normalization details.
* **Bug Fixes**
  * Freemail detection is now derived from the matched provider rule during parsing (no fixed allowlist).
* **Tests**
  * Expanded coverage for Gmail/Googlemail behavior, provider overrides, IDN/punycode matching, and quoted local-part handling.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -84,6 +88,13 @@ impl EmailAddress {
         let normalized = normalize::normalize(&parsed, config)?;
         validate::validate(&parsed, &normalized, config)?;
 
+        // Freemail status comes from the provider registry (built-ins + any
+        // custom rules), independent of provider-aware normalization.
+        let freemail = config
+            .providers
+            .lookup(&normalized.domain)
+            .is_some_and(|p| p.is_freemail());
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/lib.rs:96]: @coderabbitai Fixed — ProviderRule::new now defaults is_freemail=false; built-ins set it explicitly. Custom normalization-only rules are no longer reported as freemail.

File: src/provider.rs [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+    {
+        Self {
+            domains: domains
+                .into_iter()
+                .map(|d| d.into().to_ascii_lowercase().into_boxed_str())
+                .collect(),
+            strip_dots: false,
+            lowercase_local: false,
+            subaddress_sep: Some('+'),
+            is_freemail: true,
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/provider.rs]: Fixed — custom rules default to is_freemail=false now; built-ins opt in explicitly. The README mail.corp.example example is therefore non-freemail.

File: src/normalize.rs [hunk 3]
```diff
@@ -45,35 +45,72 @@ pub(crate) fn normalize(parsed: &Parsed<'_>, config: &Config) -> Result<Normaliz
...
 
-    // Step 2: Case folding.
-    let cased_local = match config.case_policy {
-        CasePolicy::All => nfc_local.to_lowercase(),
-        CasePolicy::Domain | CasePolicy::Preserve => nfc_local,
+    // Provider-aware overrides: when enabled and the domain matches a registered
+    // provider, that rule's case / separator / dot policy governs this address
+    // instead of the global policies. Non-matching domains use the global policies.
+    let provider = if config.provider_aware {
+        config.providers.lookup(&nfc_domain)
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/normalize.rs]: @coderabbitai Fixed — normalize() now computes the IDNA-canonical domain before provider lookup and looks up with it; freemail in parse_with already uses the canonical domain. Both call sites agree for IDN rules.

File: src/provider.rs [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+    pub fn new<I, S>(domains: I) -> Self
+    where
+        I: IntoIterator<Item = S>,
+        S: Into<String>,
+    {
+        Self {
+            domains: domains
+                .into_iter()
+                .map(|d| d.into().to_ascii_lowercase().into_boxed_str())
+                .collect(),
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/provider.rs:59]: @coderabbitai Fixed — provider domains are stored IDNA-canonicalized (domain_to_ascii) and lookups canonicalize the query, so Unicode and punycode spellings resolve identically at both call sites. Added an IDN parity test.

File: src/provider/tests.rs [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+#[test]
+fn builtin_gmail_rule() {
+    let reg = ProviderRegistry::builtin();
+    let g = reg.lookup("gmail.com").expect("gmail is a known provider");
+    assert!(g.strips_dots(), "gmail ignores dots");
+    assert!(g.folds_case());
+    assert_eq!(g.separator(), Some('+'));
+    assert!(g.is_freemail());
+    // Alias resolves to the same rule.
+    assert!(reg.lookup("googlemail.com").unwrap().strips_dots());
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/provider/tests.rs]: @coderabbitai Fixed — provider tests no longer use unwrap()/expect() (local panic helper + matches!).

File: src/normalize.rs [hunk 6]
```diff
@@ -45,53 +45,92 @@ pub(crate) fn normalize(parsed: &Parsed<'_>, config: &Config) -> Result<Normaliz
...
+        let strip = match provider {
+            Some(p) => p.strips_dots(),
+            None => match config.dot_policy {
+                DotPolicy::Preserve => false,
+                DotPolicy::Always => true,
+                // Strip only for a registered provider that ignores dots.
+                DotPolicy::GmailOnly => config
+                    .providers
+                    .lookup(&canonical_domain)
+                    .is_some_and(|p| p.strips_dots()),
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/normalize.rs]: Fixed in ef333c7. The GmailOnly branch now consults only the built-in registry (crate::provider::builtin_ref()), so custom add_provider() rules no longer leak into the legacy Gmail-only mode — custom rules apply only under provider_aware(). The built-in set moved to a process-wide LazyLock, so this hot path borrows it without a per-call allocation. Regression test added in 35bdb79 (dots_gmail_only_ignores_custom_providers).

[the remark this answers is not in the document]
Comment (polaz) [reply src/normalize.rs]: Fixed in ef333c7. The DotPolicy::GmailOnly path no longer consults config.providers; it now looks up only the built-in registry (Gmail/Googlemail). So a custom strip_dots rule from add_provider() no longer affects normalization in GmailOnly mode — custom providers take effect only under provider_aware(). Regression test in 35bdb79 covers the exact corp.example scenario you described.

File: src/provider/tests.rs [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+    // punycode matches the Unicode spelling — IDNA canonicalization is consistent.
+    let uni = ProviderRegistry::empty().with(ProviderRule::new(["münchen.de"]).freemail(true));
+    assert!(uni.lookup("münchen.de").is_some());
+    assert!(uni.lookup("xn--mnchen-3ya.de").is_some());
+
+    let puny =
+        ProviderRegistry::empty().with(ProviderRule::new(["xn--mnchen-3ya.de"]).freemail(true));
+    assert!(puny.lookup("münchen.de").is_some());
+    assert!(puny.lookup("xn--mnchen-3ya.de").is_some());
+}
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/provider/tests.rs:134]: Added in 1f60915. Two focused tests now cover the remaining Unicode edge cases on the provider lookup path: idn_rule_matches_decomposed_unicode_nfc (composed ü = U+00FC matches a decomposed u + combining diaeresis U+0308 lookup via NFC) and idn_rule_rejects_confusable_lookalike (a Cyrillic 'а' U+0430 lookalike does not match the Latin gmail.com rule). The existing test still covers Unicode/punycode aliasing.

File: src/lib.rs [hunk 8]
```diff
@@ -472,6 +455,152 @@ mod tests {
...
+        let config = Config::builder().provider_aware().build();
+        let email = EmailAddress::parse_with("\"A.B\"@gmail.com", &config)
+            .unwrap_or_else(|e| panic!("{e}"));
+        assert_eq!(email.local_part(), "A.B");
+
+        // A global lowercase policy is not provider-specific, so it still folds
+        // a quoted local-part.
+        let config = Config::builder().lowercase_all().build();
+        let email = EmailAddress::parse_with("\"A.B\"@gmail.com", &config)
+            .unwrap_or_else(|e| panic!("{e}"));
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/lib.rs:525]: @coderabbitai Done. The global-lowercase assertion now uses .provider_aware().lowercase_all(), so it exercises the regression path: the Gmail rule matches but its folding is skipped for the quoted local, while the global lowercase_all still folds it to a.b.
